### PR TITLE
Refactor func main, new TaskSet constructor

### PR DESCRIPTION
**Pull request by dontlaugh**
A new TaskSet constructor that takes functional options. The idea is that since dstask is a one-shot command line
program, we can centralize sorting, filtering, loading of tasks in the constructor of TaskSet.

Only status filtering options `WithStatuses` and `WithoutStatuses` are included in this PR. Other method-based
sorting mechanisms are retained to keep this from getting too big.

Global variables `GIT_REPO`, `STATE_FILE`, and `IDS_FILE` were removed. These are now initialized on program
startup, where we can apply overrides from configuration (such as environment variables). These are combined into a more generic `Config` type, passed down to command functions that need them, and passed down again to lower-level functions that need them. 

Refs #41

### CODE REVIEW

**Comment by dontlaugh** [comment]
@naggie This is still a WIP because I am getting the same test failure from *smoketest.sh* as the CI system. I've moved out all the commands to **commands.go**, except for the completion script generator, which is in **completions.go**.

Once I get the test passing, I'll ask for a review, since this PR is quite large as it is.

There is still more I'd like to integrate into `NewTaskSet`, such as filter options and sorting options. But that can come later.

**Comment by naggie** [comment]
Ok sounds good! Thanks.

**Comment by dontlaugh** [comment]
I'm getting closer on this, and the tests are passing. I do think there is more cleanup and simplification to do here, but it's ready for folks to take a look.

One thing we hadn't talked about before: I removed a few global variables, and that's meant passing more arguments in many files. I did this in the context of trying to sort out the failing smoke test. In retrospect, this wasn't 100% necessary, but I'd rather not back out the change.

**Comment by dontlaugh** [comment]
@naggie @ard0gg Take a look when you get a chance.

**Comment by naggie** [comment]
Generally looks good after a quick run through. I'll look in more detail tonight and I'll be using a binary from this branch for the next 2 days. Looks as though you've put a lot of effort into this, so thanks!

I've found 1 regression so far: `dstask -- show-resolved` doesn't correctly ignore the context (though `dstask show-resolved --` does). It seems the former is interpreted as the default `next` command searching for the string `show-resolved`.

File: cmd/dstask.go [hunk 1]
```diff
@@ -1,535 +1,170 @@
...
 	case "", dstask.CMD_NEXT:
-		ts := dstask.LoadTasksFromDisk(dstask.NON_RESOLVED_STATUSES)
-		ts.Filter(context)
-		ts.Filter(cmdLine)
-		ts.FilterOutStatus(dstask.STATUS_TEMPLATE)
-		ts.SortByPriority()
-		context.PrintContextDescription()
-		ts.DisplayByNext(true)
-		ts.DisplayCriticalTaskWarning()
+		if err := dstask.CommandNext(repoPath, idsFilePath, stateFilePath, ctx, cmdLine); err != nil {
```
Comment (naggie) [inline cmd/dstask.go:28]: I think it feels a bit cumbersome to pass `repoPath, idsFilePath, stateFilePath` (and perhaps even the others) for every command. I saw your comment on why you added them -- I suppose the alternatives are:

1.  Use global variables like before
2. Keep this as part of the taskset (as discussed)

What do you think?

Comment (naggie) [reply cmd/dstask.go:28]: N.B. Otherwise I like the simplification affect this is having on dstask.go.

File: completions.go [hunk 2]
```diff
@@ -0,0 +1,113 @@
+package dstask
```
Comment (naggie) [inline completions.go:1]: It's nice to have this separated!

File: const.go [hunk 3]
```diff
@@ -1,18 +1,14 @@
...
-import (
-	"os"
-	"os/user"
-	"path"
-	"strings"
-)
+import "os"
+
+func init() {
+	if os.Getenv("DSTASK_FAKE_PTY") == "1" {
```
Comment (naggie) [inline const.go:6]: I think just checking this is set to anything is best, as it might be confusing if set to some thing else (eg, `yes` is common)

File: const.go [hunk 4]
```diff
@@ -113,13 +110,13 @@ var ALL_STATUSES = []string{
...
 // incomplete until all statuses are implemented
 var VALID_STATUS_TRANSITIONS = [][]string{
-	[]string{STATUS_PENDING, STATUS_ACTIVE},
-	[]string{STATUS_ACTIVE, STATUS_PAUSED},
-	[]string{STATUS_PAUSED, STATUS_ACTIVE},
-	[]string{STATUS_PENDING, STATUS_RESOLVED},
-	[]string{STATUS_PAUSED, STATUS_RESOLVED},
-	[]string{STATUS_ACTIVE, STATUS_RESOLVED},
-	[]string{STATUS_PENDING, STATUS_TEMPLATE},
+	{STATUS_PENDING, STATUS_ACTIVE},
```
Comment (naggie) [inline const.go:113]: Hm, didn't know this was allowable.

File: git.go [hunk 5]
```diff
@@ -7,45 +7,45 @@ import (
...
-	args = append([]string{"-C", GIT_REPO}, args...)
+func RunGitCmd(repoPath string, args ...string) error {
+	args = append([]string{"-C", repoPath}, args...)
 	return RunCmd("git", args...)
 }
 
 // MustRunGitCmd delegates to RunGitCmd and exits the program on any error.
-func MustRunGitCmd(args ...string) {
-	err := RunGitCmd(args...)
+func MustRunGitCmd(repoPath string, args ...string) {
```
Comment (naggie) [inline git.go:16]: Quite a lot of variable passing again (I think I prefer the global variables / env, though not if it causes a testing issue)

File: const.go (hunk 4 above)
Comment (dontlaugh) [reply const.go:113]: Me either, but `golint` and my vim plugin were shouting at me about it.

File: cmd/dstask.go (hunk 1 above)
Comment (dontlaugh) [reply cmd/dstask.go:28]: There is another alternative that could reduce the number of parameters we take to the command functions in commands.go

We could consolidate these parameters into an application-level config struct.

```go
// config.go
type Config struct {
  Repo string `yaml:"repo"`
  IDsFile string `yaml:"ids_file"`
  StateFile string `yaml:"state_file"`
}
```

We then instantiate this on startup, using defaults as we do now so no one has to configure anything (unless they want to).

The config can be passed to the top level functions

```go
conf, _ := dstask.NewConfig(...)
if err := dstask.CommandNext(conf, ctx, cmdLine); if err != nil { /**/ }
```

For now, we could derive this struct from environment variables. But it lays the groundwork for a real config file, which I believe we will need eventually.

Of course, the lower level functions will look the same, as we will be passing values off this struct to the git commands, for instance.

Comment (naggie) [reply cmd/dstask.go:28]: That seems like a good compromise 👍

Comment (dontlaugh) [reply cmd/dstask.go:28]: I'll @ you again when the changes are up.

**Comment by dontlaugh** [comment]
I'm implemented the `Config` struct in **config.go**.

I have not yet fixed the regression related to parsing the command line after the `--`.

**Comment by dontlaugh** [comment]
This commit has a minimal test case that reproduces/fails for the regression naggie mentioned

https://github.com/naggie/dstask/pull/43/commits/f5947f07e64b5138238694e4dbabc7c279b13bd3

**Comment by dontlaugh** [comment]
@naggie This is ready for re-review, but with a caveat. I've fixed the specific regression you mentioned with updates to cmdline.go, but have identified a different one (maybe). This is in a commented out test case here.
https://github.com/naggie/dstask/pull/43/files#diff-bdaeb5f4a2b8ebcaedc9316386f14ce7

My understanding is the `IDsExhausted` boolean is only used for completions, and that this is an edge case. But, should we open an issue to handle this case?

**Comment by naggie** [comment]
`IDsExhausted` is used to stop parsing IDs -- IDs can be specified before or after a command (but not both) such that it's possible to add tasks with numbers in the description

**Comment by dontlaugh** [comment]
Ah sorry I should clarify. From my grepping of the codebase, the only part of the code that uses `IDsExhausted` outside of the parsing function is the shell completions code. So it needs to be exported for that purpose, and tested, too.

**Comment by naggie** [comment]
Oh sure I see what you mean. I guess it should be tested but we can handle that later. In the mean time I'm happy to merge, thanks for all the hard work and discussion!
